File: 01_go/sync/01_sync.go
package main

import (
	"fmt"
	"sync"
)

func drive(id int, wg *sync.WaitGroup, m *sync.Mutex, direction string) {
	m.Lock()
	if direction == "right" {
		fmt.Println("carro indo pra direita entrando na pista", id)
		fmt.Println("carro indo pra direita deixando a pista", id)
	}
	if direction == "left" {
		fmt.Println("carro indo pra esquerda entrando na pista", id)
		fmt.Println("carro indo pra esquerda deixando a pista", id)
	}
	m.Unlock()
	wg.Done()
}

func main() {
	//true -> quando há carros indo para a esquerda
	//false -> quando há carros indo para a direita
	var wg sync.WaitGroup
	var m sync.Mutex
	for i := 0; i < 100; i++ {
		wg.Add(1)
		go drive(i, &wg, &m, "left")
	}
	wg.Wait()
	fmt.Print("Acabou a fila de carros")
}

File: 01_c++/01_sync.cpp
#include <iostream>
#include <mutex>
#include <thread>

using namespace std;

mutex m;

void drive(int i,string direction){
    m.lock();
    if(direction == "left"){
        printf("carro entrando na estrada pra ir para esquerda %d\n", i);
        printf("carro indo para esquerda %d \n", i);
    }
    else if(direction == "right"){
        printf("carro entrando na estrada pra ir para direita %d\n", i);
        printf("carro indo para direita %d\n", i);
    }
    else{
        printf("Direcao Invalida\n");
    }
    m.unlock();
}

int main(){
    for(int i = 0; i < 10; i+=2){
        thread t1(drive, i, "left");
        thread t2(drive, i+1,"right");
        t1.join();
        t2.join();
    }
}
